wadas/domain/ai_model.py

```diff
@@ -170,7 +173,51 @@ def get_video_frames(self, video_path):
             yield frame, frame_count
             frame_count += 1
 
-    def process_video_offline(self, video_path):
+    def save_preview_video(self, frames, output_path):
+        """Save a sequence of PIL.Image frames as a video using OpenCV."""
+
+        if not frames:
+            raise ValueError("No frames to write to video.")
+
+        first = np.array(frames[0].convert("RGB"))
+        height, width = first.shape[:2]
+        fourcc = cv2.VideoWriter_fourcc(*"mp4v")
+        out = cv2.VideoWriter(str(output_path), fourcc, self.video_fps, (width, height))
+
+        for frame in frames:
+            rgb_array = np.array(frame.convert("RGB"))
+            bgr_array = cv2.cvtColor(rgb_array, cv2.COLOR_RGB2BGR)
+            out.write(bgr_array)
+
+        out.release()
+
+    def classification_from_video_tracking(self, tracked_animals):
+        """This method returns classification results from animal tracking out of video framing"""
+
+        # Compute average accuracy per animal id (class_name, id)
+        accuracy_data = defaultdict(list)
+
+        for frame_tracked_animals in tracked_animals:
+            for frame_tracked_animal in frame_tracked_animals:
+                class_name, accuracy = frame_tracked_animal["classification"]
+                animal_id = frame_tracked_animal["id"]
+                key = (class_name, animal_id)
+                accuracy_data[key].append(accuracy)
+
+        classified_animals = []
+        for (class_name, animal_id), accuracies in accuracy_data.items():
+            mean_accuracy = sum(accuracies) / len(accuracies)
+            classified_animals.append(
+                {
+                    "class_probs": {},
+                    "classification": [class_name, mean_accuracy],
+                    "id": animal_id,
+                    "xyxy": [],
+                }
+            )
```

Comment:
You can use a list comprehension, if you prefer:

        classified_animals = [
                {
                    "class_probs": {},
                    "classification": [class_name, sum(accuracies) / len(accuracies)],
                    "id": animal_id,
                    "xyxy": [],
                }
                for (class_name, animal_id), accuracies in accuracy_data.items()
        ]

BTW, if "classification" is not written/altered, then you can replace the [] with ().